### DISCUSSION TITLE
findFirstRecordByData throws instead of returning undefined

### DISCUSSION THREAD

**Discussion by einarpersson** [pocketbase/pocketbase#3733] [Q&A]
Hi!

Context: I am creating a invite user flow. I have a separate invites collection and send a custom email since that felt more appropriate and extensible than creating unverified users and re-purposing password-reset email.

Also: Since I would like Pocketbase to be in charge of all validation, I want pocketbase to also return sensible validation errors with codes and messages. Therefore I cannot use the admin ui API rules.

While doing this, I discovered that the code snipped below did not work: 

```ts
onRecordBeforeCreateRequest((e) => {
	const email = e.record.email()
	const inviteRecord = $app.dao().findFirstRecordByData('userInvites', 'email', email)

	if (!inviteRecord) {
		console.log('No invite found for', email)

		throw new BadRequestError('No invite found for this email address', {
			email: new ValidationError('no_invite_found', 'No invite found for this email address'),
		})
	}
}, 'users')
```

Apparently `findFirstRecordByData` throws an 404 error if not found, i.e. I need to wrap it in a try-catch instead of doing the above.

But the type signature looks like this
```ts
daos.Dao.findFirstRecordByData(collectionNameOrId: string, key: string, value: any): models.Record | undefined
// FindFirstRecordByData returns the first found record matching the provided key-value pair.
```

which hints at returning `undefined` if not found

**Suggestion:**
1. If it never returns undefined, remove it to avoid confusion.
2. Edit the method description to `FindFirstRecordByData returns the first found record matching the provided key-value pair and throws an 404 not found error otherwise.

**EDIT:** I realize that I maybe should have created an issue out of this instead of a discussion. Sorry about that!

**Comment by ganigeorgiev** [comment]
This is working as intended. All single [`find*`](https://pocketbase.io/docs/js-records/#fetch-single-record) retrieval methods should throw an error if no record matching their criteria is found.

The `undefined` in the TS return type declaration is because internally the returned value is a pointer and all Go pointer values are converted to `something | undefined` union (_the types are autogenerated from the Go source code via [pocketbase/tygoja](https://github.com/pocketbase/tygoja)_).

**Reply by ganigeorgiev** [reply]
_I'll consider in future releases to update the generated types to keep the union only when the pointer value is a function argument or struct field and make an exception for the return pointer types (at least in the PocketBase exposed methods all functions that return a Go pointer + error pair should return a valid value if the error is `nil`)._

**Reply by einarpersson** [reply]
Yes, I know this is not a bug, sorry if I wasn't clear. I have read it elsewhere. But still, it tricked me up since I saw the signature. Also, I think it could go a long way by just adding the "...or throw NotFoundError" in the method description (in intellisense).

Thanks for a great project, I enjoy working with it and only want to help make it better :)

**Comment by rainwashed** [comment]
Adding onto this year old thread, I was wondering if I can get back the original behavior of `undefined` being returned when pocketbase cannot find the resource instead of throwing an error. I want to create a user registration system and throwing errors have complicated the process of determining whether a user already exists. If it were to just return undefined, I would be able to just use that comparison and do error handling my own way.

**Reply by ganigeorgiev** [reply]
> if I can get back the original behavior of undefined being returned when

We never had this. The `findFirstRecordByData` always throw in case of a missing record. The above discussion was about its [autogenerated TS ambient declaration](https://pocketbase.io/jsvm/classes/Dao.html#findFirstRecordByData) (_it has been updated since then_).

As a workaround you can create your own helper method that wraps the call in `try/catch` but make sure to check the ["Handler scope" caveat](https://pocketbase.io/docs/js-overview/#handlers-scope).

**Reply by rainwashed** [reply]
Thanks, I will take a look into it. I haven't used PocketBase in a while, so the new system is a bit different to how I remember. I remember it used to return an undefined object, not throw an error. Thanks for the suggestion though.

**Reply by ganigeorgiev** [reply]
No, you are mistaken, there is no "new system".
